Format json error on authenticate password

<!-- Please reserve GitHub issues for bug reports and feature requests.

For questions, the best place to get answers is on our [discussion forum](https://discuss.hashicorp.com/c/boundary), as they will get more visibility from experienced users than the issue tracker.

Please note: We take Boundary's security and our users' trust very seriously. If you believe you have found a security issue in Boundary, please responsibly disclose by contacting us at security@hashicorp.com. Our PGP key is available at [our security page](https://www.hashicorp.com/security/).

-->

**Describe the bug**
Error when authenticating with boundary with -format json

**To Reproduce**
Steps to reproduce the behavior:
1. Run `boundary dev`
2. Run `boundary authenticate password -auth-method-id ampw_1234567890 -login-name admin -format json -password password`
3. See error

**Expected behavior**
To get a json response

**Additional context**
The format table works fine

This error only occurs on this boundary version
```
Version information:
  Git Revision:        c0f33f982c87c0eb4127cb16cf06b03a37b91dbd
  Version Number:      0.1.8

```

Go version
```
go version go1.16.1 linux/amd64
```

Actual error
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1244e7d]

goroutine 1 [running]:
github.com/hashicorp/boundary/api.(*Response).HttpResponse(...)
	/go/api/response.go:22
github.com/hashicorp/boundary/internal/cmd/base.(*Command).PrintJsonItem(0xc00013a000, 0x1b3bce0, 0xc000496050, 0x156ce20, 0xc000826000, 0xc000504570)
	/go/internal/cmd/base/format.go:237 +0x3d
github.com/hashicorp/boundary/internal/cmd/commands/authenticate.(*PasswordCommand).Run(0xc000504000, 0xc000138030, 0x8, 0x8, 0xc0002ec020)
	/go/internal/cmd/commands/authenticate/password.go:159 +0xeb2
github.com/mitchellh/cli.(*CLI).Run(0xc0002c6f00, 0xc0002c6f00, 0xc0002ed340, 0xc00000df80)
	/root/go/pkg/mod/github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x1cf
github.com/hashicorp/boundary/internal/cmd.RunCustom(0xc000138010, 0xa, 0xa, 0xc00039fe60, 0xc000114058)
	/go/internal/cmd/main.go:186 +0x846
github.com/hashicorp/boundary/internal/cmd.Run(...)
	/go/internal/cmd/main.go:92
main.main()
	/go/cmd/boundary/main.go:13 +0xda

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -47,6 +47,9 @@ Canonical reference for changes, improvements, and bugfixes for Boundary.
 
 * server: Roles for auto generated scopes are now generated at database init.
   [PR](https://github.com/hashicorp/boundary/pull/996)
+* cli: Don't panic on certain commands when outputting in `json` format
+  ([Issue](https://github.com/hashicorp/boundary/pull/992),
+  [PR](https://github.com/hashicorp/boundary/pull/1095))
 
 ## 0.1.8 (2021/03/10)
 

--- a/api/accounts/changepassword.go
+++ b/api/accounts/changepassword.go
@@ -57,10 +57,9 @@ func (c *Client) ChangePassword(ctx context.Context, accountId, currentPassword,
 	if err != nil {
 		return nil, fmt.Errorf("error decoding ChangePassword response: %w", err)
 	}
-
 	if apiErr != nil {
 		return nil, apiErr
 	}
-
+	target.response = resp
 	return target, nil
 }

--- a/api/accounts/setpassword.go
+++ b/api/accounts/setpassword.go
@@ -59,6 +59,6 @@ func (c *Client) SetPassword(ctx context.Context, accountId, password string, ve
 	if apiErr != nil {
 		return nil, apiErr
 	}
-
+	target.response = resp
 	return target, nil
 }

--- a/api/authmethods/authenticate.go
+++ b/api/authmethods/authenticate.go
@@ -96,6 +96,5 @@ func (c *Client) Authenticate(ctx context.Context, authMethodId, command string,
 		return nil, apiErr
 	}
 	target.response = resp
-
 	return target, nil
 }

--- a/api/authmethods/changestate.go
+++ b/api/authmethods/changestate.go
@@ -59,10 +59,9 @@ func (c *Client) ChangeState(ctx context.Context, authMethodId string, version u
 	if err != nil {
 		return nil, fmt.Errorf("error decoding ChangeState response: %w", err)
 	}
-
 	if apiErr != nil {
 		return nil, apiErr
 	}
-
+	target.response = resp
 	return target, nil
 }

--- a/api/targets/custom.go
+++ b/api/targets/custom.go
@@ -63,14 +63,15 @@ func (c *Client) AuthorizeSession(ctx context.Context, targetId string, opt ...O
 		return nil, fmt.Errorf("error performing client request during AuthorizeSession call: %w", err)
 	}
 
-	sar := new(SessionAuthorizationResult)
-	sar.Item = new(SessionAuthorization)
-	apiErr, err := resp.Decode(sar.Item)
+	target := new(SessionAuthorizationResult)
+	target.Item = new(SessionAuthorization)
+	apiErr, err := resp.Decode(target.Item)
 	if err != nil {
 		return nil, fmt.Errorf("error decoding AuthorizeSession response: %w", err)
 	}
 	if apiErr != nil {
 		return nil, apiErr
 	}
-	return sar, nil
+	target.response = resp
+	return target, nil
 }

--- a/internal/cmd/base/format.go
+++ b/internal/cmd/base/format.go
@@ -234,8 +234,10 @@ func (c *Command) PrintJsonItem(result api.GenericResult, item interface{}) bool
 		StatusCode int         `json:"status_code"`
 		Item       interface{} `json:"item"`
 	}{
-		StatusCode: result.GetResponse().HttpResponse().StatusCode,
-		Item:       item,
+		Item: item,
+	}
+	if result.GetResponse() != nil && result.GetResponse().HttpResponse() != nil {
+		output.StatusCode = result.GetResponse().HttpResponse().StatusCode
 	}
 	b, err := JsonFormatter{}.Format(output)
 	if err != nil {
@@ -252,8 +254,10 @@ func (c *Command) PrintJsonItems(result api.GenericListResult, items []interface
 		StatusCode int           `json:"status_code"`
 		Items      []interface{} `json:"items"`
 	}{
-		StatusCode: result.GetResponse().HttpResponse().StatusCode,
-		Items:      items,
+		Items: items,
+	}
+	if result.GetResponse() != nil && result.GetResponse().HttpResponse() != nil {
+		output.StatusCode = result.GetResponse().HttpResponse().StatusCode
 	}
 	b, err := JsonFormatter{}.Format(output)
 	if err != nil {

--- a/internal/cmd/commands/targetscmd/funcs.go
+++ b/internal/cmd/commands/targetscmd/funcs.go
@@ -444,11 +444,9 @@ func printCustomActionOutputImpl(c *Command) (bool, error) {
 			return true, nil
 
 		case "json":
-			b, err := base.JsonFormatter{}.Format(item)
-			if err != nil {
-				return false, fmt.Errorf("Error formatting as JSON: %w", err)
+			if ok := c.PrintJsonItem(c.sar, item); !ok {
+				return false, fmt.Errorf("Error formatting as JSON")
 			}
-			c.UI.Output(string(b))
 			return true, nil
 		}
 	}

